File: examples/s1/cpp/tutorials/src/example5_pick_and_place.cpp
#include<iostream>
#include<fstream>
#include<thread>
#include<chrono>
#include <array>
#include <cmath>
#include <algorithm>
#include <stdexcept>
#include "opencv2/opencv.hpp"

#include "galbot_robot.hpp"
#include "galbot_motion.hpp"
#include "galbot_navigation.hpp"

using namespace galbot::sdk;

inline std::vector<double> quat_normalize(const std::vector<double>& q)
{
    double norm = std::sqrt(
        q[0]*q[0] + q[1]*q[1] +
        q[2]*q[2] + q[3]*q[3]
    );
    if (norm <= 0.0)
        throw std::runtime_error("Zero-norm quaternion");

    return std::vector<double>{ q[0]/norm, q[1]/norm, q[2]/norm, q[3]/norm };
}
inline std::vector<double> quat_conjugate(const std::vector<double>& q)
{
    return std::vector<double>{ -q[0], -q[1], -q[2], q[3] };
}

inline std::vector<double> quat_multiply(const std::vector<double>& q1, const std::vector<double>& q2)
{
    const double x1 = q1[0];
    const double y1 = q1[1];
    const double z1 = q1[2];
    const double w1 = q1[3];
    const double x2 = q2[0];
    const double y2 = q2[1];
    const double z2 = q2[2];
    const double w2 = q2[3];

    return std::vector<double>{
        w1*x2 + x1*w2 + y1*z2 - z1*y2,
        w1*y2 - x1*z2 + y1*w2 + z1*x2,
        w1*z2 + x1*y2 - y1*x2 + z1*w2,
        w1*w2 - x1*x2 - y1*y2 - z1*z2
    };
}

template <typename T>
inline T clamp(const T& v, const T& lo, const T& hi)
{
    return (v < lo) ? lo : (v > hi) ? hi : v;
}

inline double orientation_error_angle(const std::vector<double>& A, const std::vector<double>& B)
{
    std::vector<double> qA = quat_normalize({ A[3], A[4], A[5], A[6] });
    std::vector<double> qB = quat_normalize({ B[3], B[4], B[5], B[6] });

    std::vector<double> q_err = quat_multiply(qB, quat_conjugate(qA));
    q_err = quat_normalize(q_err);

    // Numerically stable
    double qw = clamp(q_err[3], -1.0, 1.0);

    return 2.0 * std::acos(qw);
}

struct PoseError
{
    double position_error_norm;   // meters
    double orientation_error_rad;
    double orientation_error_deg;
};

inline PoseError calculate_error(const std::vector<double>& pose1, const std::vector<double>& pose2)
{
    double dx = pose1[0] - pose2[0];
    double dy = pose1[1] - pose2[1];
    double dz = pose1[2] - pose2[2];

    double pos_err = std::sqrt(dx*dx + dy*dy + dz*dz);
    double rot_err = orientation_error_angle(pose1, pose2);

    return PoseError{
        pos_err,
        rot_err,
        rot_err * 180.0 / M_PI
    };
}


/* @brief Check if the robot is safe
*/
void check_robot_safety(){
    std::cout << "⚠️  Note: 1. Please ensure the robot's emergency stop button is released; 2. Please ensure there are no obstacles in front, back, left, and right of the robot to avoid unexpected situations. \n" << std::endl;

    char key;
    for(;;){
        std::cout << "Please confirm that the robot's emergency stop button is released and there are no obstacles. Continue? (y/n)...";
        std::cin >> key;

        if(std::tolower(key) == 'y'){
            std::cout << "User confirmed, continuing execution...\n" << std::endl;
            break;
        }else if(std::tolower(key) == 'n'){
            std::cout << "User not confirmed, program exiting...\n" << std::endl;
            exit(0);
        }else{
            std::cout << "Input error, please enter 'y' or 'n'\n" << std::endl;
        }
    }
}

Pose get_navigation_pose(const Pose& object_pose, const GalbotMotion& motion){
    /**
     * Please complete the function to calculate the navigation pose.
     * For convenience, we set it default as [0.7, -0.3, 0.0, 0.0, 0.0, 0.0, 1.0],
     */

    return Pose(std::vector<double>{0.7, 0.7, 0.0, 0.0, 0.0, 0.0, 1.0});
}

void navigation_to_goal(GalbotNavigation& navi, const Pose& goal_pose){
    Pose current_pose = navi.get_current_pose();
    std::cout << "Current pose: [" 
              << current_pose.position.x << ", "
              << current_pose.position.y << ", "
              << current_pose.position.z << "]" << std::endl;

    if (navi.check_path_reachability(goal_pose, current_pose)){
        int retry_cnt = 3;
        for(;;){
            NavigationStatus status = navi.navigate_to_goal(goal_pose, true, true, 20.0);
            std::this_thread::sleep_for(std::chrono::seconds(1));
            retry_cnt -= 1;
            if (navi.check_goal_arrival() || retry_cnt < 0){
                break;
            }else{
                std::cout << "Navigation failed, retry count: " << retry_cnt << std::endl;
            }
        }
        std::cout << "has arrived: " << navi.check_goal_arrival() << std::endl;
    }else{
        std::cout << "The path to the goal is not reachable." << std::endl;
    }
}

void lift_camera_up(GalbotMotion& motion, const std::string& target_chain, double lift_delta_z = 0.05){
    int retry_cnt = 3;
    std::vector<double> cur_ee_pose;
    std::string frame_id = "EndEffector";
    std::string reference_frame = "base_link";
    for(;;){
        std::tuple<MotionStatus, std::vector<double>> motion_status = motion.get_end_effector_pose_on_chain(
            target_chain, frame_id, reference_frame
        );
        std::this_thread::sleep_for(std::chrono::seconds(1));

        retry_cnt -= 1;
        if (retry_cnt < 0){
            std::cout << "Retry " << retry_cnt << " times, get end effector pose failed." << std::endl;
            break;
        }else if (std::get<0>(motion_status) == MotionStatus::SUCCESS){
            cur_ee_pose = std::get<1>(motion_status);
            std::cout << "Get end effector pose success, pose: [" 
                      << cur_ee_pose[0] << ", "
                      << cur_ee_pose[1] << ", "
                      << cur_ee_pose[2] << "]" << std::endl;
            break;
        }else{
            std::cout << "Get end effector pose failed, retry count: " << retry_cnt << std::endl;
        }
    }

    std::vector<double> target_ee_pose = cur_ee_pose;
    target_ee_pose[2] += lift_delta_z;

    retry_cnt = 3;
    for(;;){
        MotionStatus motion_status2 = motion.set_end_effector_pose(
            target_ee_pose,
            target_chain, 
            reference_frame,
            nullptr,
            false, 
            true, 
            5.0,
            std::make_shared<Parameter>()
        );
        std::this_thread::sleep_for(std::chrono::seconds(1));

        retry_cnt -= 1;
        if (retry_cnt < 0){
            std::cout << "Retry " << retry_cnt << " times, set end effector pose failed." << std::endl;
            break;
        }else if (motion_status2 == MotionStatus::SUCCESS){
            std::cout << "Set end effector pose success." << std::endl;
            break;
        }else{
            std::cout << "Set end effector pose failed, retry count: " << retry_cnt << std::endl;
        }
    }
}

std::vector<double> detect_target(std::shared_ptr<cv::Mat> rgb_image, std::shared_ptr<cv::Mat> depth_image){
    /**
     * This function is a placeholder. In a real-world scenario, you would implement
     * target detection using computer vision techniques. For this example, we assume a default pose.
     */
    return std::vector<double>{-0.05, 0.1, 0.12, 0.0, 0.0, 0.0, 1.0};
}

std::vector<double> pose_camera_to_base(GalbotRobot& robot, const Pose& object_pose_camera){
    /**
     * This function is a placeholder. You should implement the transformation from camera pose to chassis pose.
     */
    return std::vector<double>{0.2, 0.2, 0.8, 0.0, 0.0, 0.0, 1.0};
}

std::vector<double> detect_object(GalbotRobot& robot, std::string arm){
    /** Get rgb and depth data from robot sensor */
    std::shared_ptr<RgbData> rgb_data;
    std::shared_ptr<DepthData> depth_data;
    if (arm == "left_arm"){
        rgb_data = robot.get_rgb_data(SensorType::LEFT_ARM_CAMERA);
        depth_data = robot.get_depth_data(SensorType::LEFT_ARM_DEPTH_CAMERA);
    }else if (arm == "right_arm"){
        rgb_data = robot.get_rgb_data(SensorType::RIGHT_ARM_CAMERA);
        depth_data = robot.get_depth_data(SensorType::RIGHT_ARM_DEPTH_CAMERA);
    }else{
        std::cout << "Invalid arm name: " << arm << std::endl;
    }

    /** Convert rgb and depth data to cv::Mat */
    std::shared_ptr<cv::Mat> rgb_image, depth_image;
    if (rgb_data){
        rgb_image = rgb_data->convert_to_cv2_mat();
        std::cout << "Get rgb image success" << std::endl;
    }else{
        std::cout << "Get rgb image failed" << std::endl;
    }
    if (depth_data){
        depth_image = depth_data->convert_to_cv2_mat();
        std::cout << "Get depth image success" << std::endl;
    }else{
        std::cout << "Get depth image failed" << std::endl;
    }

    if (rgb_image && depth_image){
        std::cout << "Detected target..." << std::endl;
        std::vector<double> object_pose_camera = detect_target(rgb_image, depth_image);
        std::vector<double> object_pose_base = pose_camera_to_base(robot, object_pose_camera);
        return object_pose_base;
    }else{
        std::cout << "Get rgb or depth image failed" << std::endl;
        return std::vector<double>{0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0};
    }
}

void pick_and_place(
    GalbotRobot& robot, 
    GalbotMotion& motion, 
    GalbotNavigation& navi, 
    const std::vector<double>& object_pose_base, 
    std::string target_chain, 
    std::string reference_frame){
    
    /** Open gripper */
    ControlStatus status = robot.set_gripper_command(
        "left_gripper", 0.1, 0.05, 10, false
    );
    std::this_thread::sleep_for(std::chrono::seconds(1));
    if (status == ControlStatus::SUCCESS){
        std::cout << "Open gripper success" << std::endl;
    }else{
        std::cout << "Open gripper failed" << std::endl;
    }

    /** Move to pre-grasp pose */
    int retry_cnt = 3;
    for(;;){
        MotionStatus motion_status = motion.set_end_effector_pose(
            object_pose_base,
            target_chain, 
            reference_frame,
            nullptr,
            false, 
            true, 
            5.0,
            std::make_shared<Parameter>()
        );
        std::this_thread::sleep_for(std::chrono::seconds(1));

        retry_cnt -= 1;
        if (retry_cnt < 0){
            std::cout << "Retry " << retry_cnt << " times, set end effector pose failed." << std::endl;
            break;
        }else if (motion_status == MotionStatus::SUCCESS){
            std::cout << "Set end effector pose success." << std::endl;
            break;
        }else{
            std::cout << "Set end effector pose failed, retry count: " << retry_cnt << std::endl;
        }
    }

    /** Close gripper */
    ControlStatus status1 = robot.set_gripper_command(
        "left_gripper", 0.02, 0.05, 10, false
    );
    std::this_thread::sleep_for(std::chrono::seconds(1));
    if (status1 == ControlStatus::SUCCESS){
        std::cout << "Close gripper success" << std::endl;
    }else{
        std::cout << "Close gripper failed" << std::endl;
    }

    /** Return to initial position */
    navigation_to_goal(navi, Pose(std::vector<double>{0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0}));
    std::this_thread::sleep_for(std::chrono::seconds(1));

    /** Open gripper to release object */
    ControlStatus status2 = robot.set_gripper_command(
        "left_gripper", 0.1, 0.05, 10, false
    );
    std::this_thread::sleep_for(std::chrono::seconds(1));
    if (status2 == ControlStatus::SUCCESS){
        std::cout << "Open gripper to release object success" << std::endl;
    }else{
        std::cout << "Open gripper to release object failed" << std::endl;
    }
}

int main(){
    check_robot_safety();
    try{
        /* Get robot instance  */
        auto& robot = GalbotRobot::get_instance(MachineType::S1); 
        auto& motion = GalbotMotion::get_instance(MachineType::S1);
        auto& navi = GalbotNavigation::get_instance(MachineType::S1);        

        /* Initialize robot */
        if (robot.init({SensorType::LEFT_ARM_CAMERA, SensorType::LEFT_ARM_DEPTH_CAMERA})) {
            std::cout << "Initialization successful" << std::endl;
            std::cout << "Is robot running: " << robot.is_running() << std::endl;
        }else{
            std::cerr << "Initialization failed" << std::endl;
        }
        /** Initialize motion */
        if (motion.init()) {
            std::cout << "Motion initialization successful" << std::endl;
        }else{
            std::cerr << "Motion initialization failed" << std::endl;
        }
        /** Initialize navigation */
        if (navi.init()) {
            std::cout << "Navigation initialization successful" << std::endl;
        }else{
            std::cerr << "Navigation initialization failed" << std::endl;
        }

        /* Wait for data preparation */
        std::this_thread::sleep_for(std::chrono::milliseconds(3000));
        
        /** Calculate navigation target pose */
        Pose object_goal_pose(std::vector<double>{1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 1.0});
        Pose base_goal_pose = get_navigation_pose(object_goal_pose, motion);

        /** Navigate to the navigation pose */
        navigation_to_goal(navi, base_goal_pose);

        /** Get original joint positions for recovery */
        std::string target_chain = "left_arm";
        std::string reference_frame = "base_link";
        std::vector<double> original_joint_positions = robot.get_joint_positions(std::vector<std::string>{target_chain}, {});

        /** Lift the camera up for better view */
        lift_camera_up(motion, target_chain);

        /** Detect object pose */
        std::vector<double> object_pose_base = detect_object(robot, target_chain);

        /** Pick the object and move to original position */
        pick_and_place(robot, motion, navi, object_pose_base, target_chain, reference_frame);
        
        ControlStatus status = robot.set_joint_positions(
            original_joint_positions, 
            std::vector<std::string>{target_chain}, 
            {},
            true,
            0.3,
            30.0
        );
        if (status == ControlStatus::SUCCESS) {
            std::cout << "Joint recover sucess." << std::endl;
        } else {
            std::cerr << "Joint recover failed." << std::endl;
        }

        /** Actively send SIGINT exit signal to the robot */
        robot.request_shutdown();
        /** Wait to enter shutdown state */
        robot.wait_for_shutdown();
        /** Release SDK resources */
        robot.destroy();
        std::cout << "Resource release successful" << std::endl;    
    }catch(const std::exception& e){
        std::cout << "Error: " << e.what() << std::endl;
    }

    return 0;
}

File: docs/g1/zh/python/tutorials/example5_pick_and_place.py
"""
Note: When running this example, please confirm that the robot's motion control service `/data/galbot/bin/service_motion_plan`,
    robot state publishing service `/data/galbot/bin/robot_state_publish`,
    navigation service `/data/galbot/bin/service_navigation_plan`
    and hand-eye calibration publishing service `/data/galbot/bin/eyehand_calib_publish` have been loaded;
"""
try:
    import galbot_sdk.g1 as gm
    from galbot_sdk.g1 import GalbotNavigation
    from galbot_sdk.g1 import GalbotRobot
    from galbot_sdk.g1 import GalbotMotion
    from galbot_sdk.g1 import G1JointGroup, SensorType
    from galbot_sdk.g1 import ControlStatus
except ImportError:
    print("Import galbot_sdk failed, please install it first or check if it is in the PYTHONPATH")
    exit(1)

import os

try:
    import numpy as np
except ImportError:
    os.system("pip install numpy")
    import numpy as np

try:
    from scipy.spatial.transform import Rotation as R
except ImportError:
    os.system("pip install scipy")
    from scipy.spatial.transform import Rotation as R

try:
    import cv2
except ImportError:
    os.system("pip install opencv-python")
    import cv2

import time
from typing import Sequence

def robot_default(robot: GalbotRobot):
    # set initial joint positions
    joint_pos = [0.5, 1.5, 1.0, 0.0, 0.0,
                    0.0, 0.0,
                    2.0, -1.5, -0.6, -1.7, 0.0, -0.8, 0.0,
                    -2.0, 1.5, 0.6, 1.7, 0.0, 0.8, 0.0]
    joint_groups_names = ["leg", "head", "left_arm", "right_arm"]
    joint_names = []
    is_blocking = True
    max_speed_rad_s = 0.1
    timeout_s = 30.0

    status = robot.set_joint_positions(
        joint_pos, joint_groups_names, joint_names, is_blocking, max_speed_rad_s, timeout_s
    )

    if status != ControlStatus.SUCCESS:
        print("set join position failed")
    else:
        print("set join position successful")

def decode_compressed_image(compressed_image, camera_info={}):
    """
    decode CompressedImage image

    Parameters:
        compressed_image (dict): image dict, keys:[header, format, data, "depth_scale"]

    Returns:
        numpy.ndarray: decoded image
    """
    image_data = compressed_image["data"]
    if compressed_image["format"] == "rgb8":
        return decode_rgb_image(image_data)
    elif compressed_image["format"] == "16UC1":
        return decode_depth_image(image_data, compressed_image["depth_scale"], camera_info)
    else:
        raise ValueError(f"Unsupport data format: {compressed_image['format']}")

def decode_rgb_image(image_data):
    """decode rgb image"""
    nparr = np.frombuffer(image_data, np.uint8)
    img = cv2.imdecode(nparr, cv2.IMREAD_COLOR)
    if img is None:
        raise ValueError("Fail to Decode RGB Image")
    return img

def decode_depth_image(image_data, depth_scale, camera_info):
    """decode depth image"""
    depth_img = np.frombuffer(image_data, dtype=np.uint16).copy()

    if not camera_info:
        depth_img = depth_img.reshape((720, 1280))
    else:
        depth_img = depth_img.reshape((camera_info["height"], camera_info["width"]))
    depth_img = depth_img.astype(np.float32) / depth_scale

    return depth_img

def print_gripper_state(joint_group, gripper_state):
    """
    Print gripper state

    Parameters:
        joint_group (G1JointGroup): G1JointGroup enum
        gripper_state (object): Contains timestamp_ns, width, velocity, effort, is_moving
    """
    print(f"Timestamp (ns): {gripper_state.timestamp_ns}")
    print(
        f"width {gripper_state.width} "
        f"velocity {gripper_state.velocity} "
        f"effort {gripper_state.effort} "
        f"is moving {gripper_state.is_moving}"
    )

def get_navigation_pose(object_goal_pose: Sequence[float], motion: GalbotMotion, arm: str = "left_arm"):
    """
    Get navigation target pose

    Parameters:
        object_goal_pose (Sequence[float]): Target pose [x, y, z, qx, qy, qz, qw]
        motion (GalbotMotion): Motion control instance
        arm (str, optional): End effector name. Defaults to "left_arm".

    Returns:
        Sequence[float]: Navigation target pose [x, y, z, qx, qy, qz, qw]
    """
    assert arm in ["left_arm", "right_arm"], "arm must be left_arm or right_arm"

    try:
        status, ee_pose_in_base = motion.get_end_effector_pose(
            end_effector_frame=f"{arm}_end_effector_mount_link",
            reference_frame="base_link"
        )
        if status != gm.MotionStatus.SUCCESS:
            print(f"Failed to get end effector pose: status={status}")
            offset_y = ee_pose_in_base[1]
        else:
            print(f"Successfully got end effector pose: pose={ee_pose_in_base}")
            offset_y = 0.3

        # Set chassis pose to the same z coordinate as the target pose
        base_goal_pose_mat = np.eye(4)
        base_goal_pose_mat[:3, :3] = R.from_quat(object_goal_pose[3:]).as_matrix()
        base_goal_pose_mat[:3, 3] = np.array([object_goal_pose[0], object_goal_pose[1], 0])

        # According to the relative position of the chassis and camera, move the camera navigation target 0.6m backward in the local coordinate to leave observation space for the camera
        base_goal_pose_mat = base_goal_pose_mat @ np.array([[1,0,0,-0.6],[0,1,0,-offset_y],[0,0,1,0],[0,0,0,1]])
        print(base_goal_pose_mat)

        base_goal_pose_quat = R.from_matrix(base_goal_pose_mat[:3, :3]).as_quat()
        base_goal_pose_pos = base_goal_pose_mat[:3, 3]

        return base_goal_pose_pos.tolist() + base_goal_pose_quat.tolist()
    
    except Exception as e:
        print("Failed to get navigation target pose:", e)
    

def navigation_to_goal(nav: GalbotNavigation, goal_pose: Sequence[float], retry_cnt: int = 3):
    """
    Navigate to target pose

    Parameters:
        nav (GalbotNavigation): Navigation instance
        goal_pose (Sequence[float]): Target pose [x, y, z, qx, qy, qz, qw]
        retry_cnt (int, optional): Number of retries. Defaults to 3.
    """
    try:
        cur_pose = nav.get_current_pose()
        print(f"Current pose: {cur_pose}")
        if nav.check_path_reachability(goal_pose, cur_pose):
            retry_cnt = 3
            while True:
                status = nav.navigate_to_goal(goal_pose, enable_collision_check=True, is_blocking=True, timeout=20)
                time.sleep(0.5)
                retry_cnt -= 1
                if nav.check_goal_arrival() or retry_cnt < 0:
                    break
                else:
                    print(f"Navigation failed: status={status}, retrying: {retry_cnt}")
            print("navigate_to_goal return status:", status)
            print("Has arrived:", nav.check_goal_arrival())
        else:
            print("Path unreachable or unsafe")
    except Exception as e:
        print(f"Exception occurred during navigation: {e}")

def lift_camera_up(motion: GalbotMotion, target_pose: Sequence[float], target_chain: str, reference_frame: str):
    """
    Lift camera to target height

    Parameters:
        motion (GalbotMotion): Motion control instance
        target_pose (Sequence[float]): Target pose [x, y, z, qx, qy, qz, qw]
        target_chain (str): End effector name
        reference_frame (str): Reference frame
    """
    try:
        retry_cnt = 3
        while True:
            status, cur_ee_pose = motion.get_end_effector_pose_on_chain(
                chain_name=target_chain,
                frame_id="EndEffector",
                reference_frame=reference_frame
            )
            time.sleep(0.5)
            retry_cnt -= 1

            if status == gm.MotionStatus.SUCCESS or retry_cnt < 0:
                print(f"Current end effector pose: {cur_ee_pose}")
                break
            else:
                print(f"Failed to get end effector pose: status={status}, retrying: {retry_cnt}")
        
        tgt_ee_pose = cur_ee_pose.copy()
        tgt_ee_pose[2] = target_pose[2] - 0.1
        print(f"Target end effector pose: {tgt_ee_pose}")

        retry_cnt = 3
        while True:
            status = motion.set_end_effector_pose(
                target_pose=tgt_ee_pose,
                end_effector_frame=target_chain,
                reference_frame=reference_frame,
                enable_collision_check=False,
                is_blocking=True,
                timeout=5.0,
                params=gm.Parameter()
            )
            time.sleep(0.5)
            retry_cnt -= 1
            if status == gm.MotionStatus.SUCCESS or retry_cnt < 0:
                print(f"✅ Successfully set end effector pose: status={status}")
                break
            else:
                print(f"Failed to set end effector pose: status={status}, retrying: {retry_cnt}")
    except Exception as e:
        print(f"❌ Failed to lift camera: {e}")

def detect_target(img: np.ndarray, depth_img: np.ndarray) -> Sequence[float]:
    """
    Detection target function. Input RGB image and depth image, output target pose.

    Parameters:
        img (np.ndarray): RGB image
        depth_img (np.ndarray): Depth image

    Returns:
        Sequence[float]: Target pose [x, y, z, qx, qy, qz, qw]
    """
    try:
        
        ############### NOTE ###############
        # This function is a placeholder. In a real-world scenario, you would implement
        # target detection using computer vision techniques. For this example, we assume
        # a default pose.
        ####################################

        # Assume detected target pose is [-0.05, -0.1, 0.12, 0.0, 0.0, 0.0, 1.0]
        # Indicates the target is 0.12m in front of the camera, 0.05m to the left, 0.1m in height, facing the default camera direction
        default_pose = [-0.05, -0.1, 0.12, 0.0, 0.0, 0.0, 1.0]
        
        return default_pose
    except Exception as e:
        print(f"Target detection exception: {e}")
        return None

def pose_camera_to_base(robot: GalbotRobot, pose_camera: Sequence[float]) -> Sequence[float]:
    """
    Transform camera pose to chassis coordinate system

    Parameters:
        robot (GalbotRobot): Robot instance
        pose_camera (Sequence[float]): Camera pose [x, y, z, qx, qy, qz, qw]

    Returns:
        Sequence[float]: Chassis pose [x, y, z, qx, qy, qz, qw]
    """
    source_frame="left_arm_camera_color_optical_frame"
    target_frame="base_link"
    base_to_cam = robot.get_transform(target_frame, source_frame)[0]
    if base_to_cam is None:
        print("Failed to get transform from camera to chassis")
        return None
    else:
        print("base_to_cam: ", base_to_cam)

    base_to_cam_mat = np.eye(4)
    base_to_cam_mat[:3, :3] = R.from_quat(base_to_cam[3:]).as_matrix()
    base_to_cam_mat[:3, 3] = np.array(base_to_cam[:3])
    
    pose_camera_mat = np.eye(4)
    pose_camera_mat[:3, :3] = R.from_quat(pose_camera[3:]).as_matrix()
    pose_camera_mat[:3, 3] = np.array(pose_camera[:3])
    
    pose_base_mat = base_to_cam_mat @ pose_camera_mat[:, 3:]
    print("pose_base_mat: ", pose_base_mat)
    
    return pose_base_mat.flatten()[:3].tolist() + [0, 0, 0, 1]

def detect_object(robot: GalbotRobot, arm: str = "left_arm"):
    try:
        # Get camera image data
        if arm == "left_arm":
            rgb_image_data = robot.get_rgb_data(SensorType.LEFT_ARM_CAMERA)
            depth_data = robot.get_depth_data(SensorType.LEFT_ARM_DEPTH_CAMERA)
        elif arm == "right_arm":
            rgb_image_data = robot.get_rgb_data(SensorType.RIGHT_ARM_CAMERA)
            depth_data = robot.get_depth_data(SensorType.RIGHT_ARM_DEPTH_CAMERA)
        else:
            raise ValueError("arm must be left_arm or right_arm")

        # Decode image data
        if not rgb_image_data:
            print("No rgb image data!")
        else:
            print("get rgb image suceess")
            img = decode_compressed_image(rgb_image_data)
        
        if not depth_data:
            print("No depth_data!")
        else:
            depth_img = decode_compressed_image(depth_data)
            print("get depth data suceess")

        # Detect target
        object_pose_camera = detect_target(img, depth_img)
        if object_pose_camera is None:
            print("Target detection failed")
            return None
        else:
            print(f"object_pose_camera: {object_pose_camera}")
        
        # Calculate target pose in chassis coordinate system
        object_pose_base = pose_camera_to_base(robot, object_pose_camera)
        print(f"Target pose in chassis coordinate system: {object_pose_base}")
    
    except Exception as e:
        print(f"Target detection exception: {e}")

    return object_pose_base

def check_robot_safety():
    """Check if robot is safe"""
    # Prompt important notes
    print("⚠️  Note: 1. Please ensure the emergency stop button of the robot is released; 2. Please ensure there are no obstructions around the robot to avoid unexpected situations. 3. Please ensure the area around the robot is clear of obstacles.")
    while True:
        key = input("Please confirm that the robot's emergency stop button is released and there are no obstructions, continue? (y/n)...")
        if key == 'y':
            print("User confirmed, continuing...")
            break
        elif key == 'n':
            print("User did not confirm, exiting program...")
            exit(1)
        else:
            print("Invalid input, please enter 'y' or 'n'")

def pick_and_place(robot: GalbotRobot, 
                   nav: GalbotNavigation, 
                   motion: GalbotMotion, 
                   object_pose_base: Sequence[float], 
                   target_chain: str, 
                   reference_frame: str):
    try:
        # Open left gripper
        # Set left gripper width to 0.1m, speed to 0.05m, force to 10N, will block until gripper reaches position
        status = robot.set_gripper_command(
            G1JointGroup.left_gripper, 0.1, 0.05, 10, False
        )
        time.sleep(0.5)

        print("object_pose_base: ", object_pose_base)
        
        # Reach to target position
        retry_cnt = 3
        while True:
            status = motion.set_end_effector_pose(
                target_pose=object_pose_base,
                end_effector_frame=target_chain,
                reference_frame=reference_frame,
                enable_collision_check=False,
                is_blocking=True,
                timeout=5.0,
                params=gm.Parameter()
            )
            time.sleep(1)
            retry_cnt -= 1
            if status == gm.MotionStatus.SUCCESS or retry_cnt < 0:
                break
            else:
                print(f"Failed to set end effector pose: status={status}, retry count: {retry_cnt}")
        
        assert status == gm.MotionStatus.SUCCESS, "Failed to set end effector pose"
        print(f"✅ Successfully set end effector pose: status={status}")

        # Close gripper to grasp object
        status = robot.set_gripper_command(
            G1JointGroup.left_gripper, 0.02, 0.05, 10, False
        )
        time.sleep(0.5)

        # Return to initial position
        navigation_to_goal(nav, [0, 0, 0, 0, 0, 0, 1])
        time.sleep(2)

        # Release target
        status = robot.set_gripper_command(
            G1JointGroup.left_gripper, 0.1, 0.05, 10, False
        )
        time.sleep(0.5)
    except Exception as e:
        print(f"Exception occurred during pick_and_place: {e}")
        return None


def main():
    check_robot_safety()
    try:
        # Get robot instance
        robot = GalbotRobot()
        # Get GalbotMotion instance
        motion = GalbotMotion()
        # Get navigation instance
        nav = GalbotNavigation()

        # Get RGB and depth images from left arm, depth images from right arm, 
        # base LiDAR data, and torso IMU data
        enable_sensor_set = {SensorType.LEFT_ARM_CAMERA, # Left arm depth camera
                            SensorType.LEFT_ARM_DEPTH_CAMERA, # Left arm RGB camera
                            SensorType.BASE_LIDAR, # Base LiDAR
                            SensorType.TORSO_IMU} # Torso IMU sensor

        # Initialize robot
        if robot.init(enable_sensor_set):
            print("GalbotRobot initialization successful")
        else:
            print("GalbotRobot initialization failed")
        if motion.init():
            print("GalbotMotion initialization successful")
        else:
            print("GalbotMotion initialization failed")
        if nav.init():
            print("GalbotNavigation initialization successful")
        else:
            print("GalbotNavigation initialization failed")

        # Program starts immediately, wait for data readiness
        time.sleep(1)

        robot_default(robot);

        # Calculate navigation target pose
        object_goal_pose = [-1, 0.33, 0.90, 0, 0, 1, 0]
        base_goal_pose = [0, 0, 0, 0, 0, 0, 1]
        base_goal_pose = get_navigation_pose(object_goal_pose, motion)
        
        # Navigate to target pose
        navigation_to_goal(nav, base_goal_pose)
        
        target_chain = "left_arm"
        reference_frame = "base_link"

        # Get current end effector pose for subsequent restoration
        try:
            status, original_pose = motion.get_end_effector_pose_on_chain(
                chain_name=target_chain,
                frame_id="EndEffector",
                reference_frame=reference_frame
            )
            assert status == gm.MotionStatus.SUCCESS, "Failed to get end effector pose"
            print(f"✅ Successfully got {target_chain} end effector pose: {original_pose}")
            time.sleep(1)
        except Exception as e:
            print(f"❌ {target_chain} end effector pose exception: {e}")

        # Lift camera
        lift_camera_up(motion, object_goal_pose, "left_arm", "base_link")

        # Detect target
        object_pose_base = detect_object(robot, arm="left_arm")
        if object_pose_base is None:
            print("Target detection failed")
            return None
        else:
            print(f"Detected target pose: {object_pose_base}")
        
        # Grasp and return to initial position
        pick_and_place(robot, nav, motion, object_pose_base, "left_arm", "base_link")

        # After grasping target, restore posture
        try:
            time.sleep(2)
            status = motion.set_end_effector_pose(
                target_pose=original_pose,
                end_effector_frame=target_chain,
                reference_frame=reference_frame,
                enable_collision_check=False,
                is_blocking=True,
                timeout=5.0,
                params=gm.Parameter()
            )
            assert status == gm.MotionStatus.SUCCESS, "Failed to set end effector pose"
            print(f"✅ Successfully set end effector pose: status={status}")
            time.sleep(1)
        except Exception as e:
            print(f"❌ {target_chain} end effector pose exception: {e}")
            
    except Exception as e:
        print(f"Exception occurred: {e}")
    finally:
        # Actively send SIGINT exit signal
        robot.request_shutdown()
        # Wait to enter shutdown state
        robot.wait_for_shutdown()
        # Release SDK resources
        robot.destroy()
        print('Resource release successful')

if __name__ == "__main__":
    main()